tests/functional_tests/util_resources.py

```diff
@@ -0,0 +1,117 @@
+#    prior written permission.
+# 
+# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY
+# EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
+# MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL
+# THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
+# PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+# INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT,
+# STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF
+# THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+"""
+    Help determine how much CPU power is available at the given time
+    by running numerical calculations
+"""
+
+from __future__ import print_function
+import datetime
+import multiprocessing
+import psutil
+from math import pi
+
+def available_ram_gb():
+    ram_bytes = psutil.virtual_memory().available
+    kilo = 1024
+    ram_gb = ram_bytes / kilo / kilo / kilo
+    return ram_gb
+
+def pi_digits_lib(digits):
+    """
+    Returns the number of digits of Pi
+    """
+    return float('{:.{}f}'.format(pi, digits))
+
+def pi_digits_numeric_yield(digits):
+    """Generate x digits of Pi.
+    https://stackoverflow.com/a/54967370
+    """
+    # Make the format the same as the library call
+    # here we need 2 chars more for '3' and '.'
+    x = digits + 2
+    k,a,b,a1,b1 = 2,4,1,12,4
+    while x > 0:
+        p,q,k = k * k, 2 * k + 1, k + 1
+        a,b,a1,b1 = a1, b1, p*a + q*a1, p*b + q*b1
+        d,d1 = a/b, a1/b1
+        while d == d1 and x > 0:
+            yield int(d)
+            x -= 1
+            a,a1 = 10*(a % b), 10*(a1 % b1)
+            d,d1 = a/b, a1/b1
+
+
+def pi_digits_numeric(digits):
+    "Get PI rounded to n digits"
+    my_array = []
+    for i in pi_digits_numeric_yield(digits):
+        my_array.append(str(i))
+        
+    first = my_array.pop(0) # Pop the 1st 3
+    valStr = "{}.{}".format(first, "".join(my_array))
+    val = float(valStr)
+    rounded = round(val, digits)
+    
+    # Compare if the calcualted Pi is the same as the library one
+    pi_library = pi_digits_lib(digits)
+    if pi_library != rounded:
+        raise RuntimeError("Calculated Pi = ", rounded, "Library Pi = ", pi_library)
+    return rounded
+        
+def get_time_pi(num_digits, cores=multiprocessing.cpu_count()):
+    """   
+    Get the time of generation of pi, using the chosen number of cores.
+    This helps in estimating how well the problem scales on multiple cores.
+    """
+    # Inputs are the number of digits to calculate for each CPU
+    inp = [num_digits] * cores
+    # Start measuring the time
+    start = datetime.datetime.now()
```

Comment:
The methodology here ends up counting process startup time, which could lead to wildly inaccurate results. A simple workaround would be to have each task time itself and push results onto a Queue.

Reply:
Yes, it does count the process start up time, but if it increases by a large margin, so will an equivalent functionality of the RX init just in the next 2 seconds after this calculation.

Reply:
Well, first off, it's not accurate to assume process startup time will be relatively constant over a short timeframe, but second, even if one makes such an assumption, I fail to see how it resolves my concern.

Reply:
You are right from the logical PoV. I would just like to cover the startup up with a calculation long enough, that the process init becomes negligible. I have a small problem with your approach, that it makes a simple thing slightly more complex. I wanted to solve this with averaging the returned processing time of the threads. I only fear, that the time calculation within the threads will be not accurate. Do you think the monotonic timer will help?

Reply:
I think running a benchmark with a broken methodology makes everything more complex, as it's very difficult to reason about a calculation with many different noise sources (are they acceptable? did the code intentionally leave them in play, and if so, why? etc). Using a Queue maybe adds 5-10 LoC and much clarity. As for timing process startup, I proposed just not even bothering and assuming some fixed value, so that's not really a concern, I think (you could easily time it, too, but I don't have a strong opinion on the matter).